src/main/java/org/ow2/proactive/connector/iaas/cloud/provider/jclouds/aws/AWSEC2JCloudsProvider.java

```diff
@@ -222,4 +242,29 @@ public void removeInstancePublicIp(Infrastructure infrastructure, String instanc
         });
     }
 
+    @Override
+    public String createKeyPair(Infrastructure infrastructure, Instance instance) {
+        KeyPairApi keyPairApi = getKeyPairApi(infrastructure);
+        String region = getRegionFromImage(instance);
+        String keyPairName = infrastructure.getId();
+        KeyPair keyPair = keyPairApi.createKeyPairInRegion(region, keyPairName);
```

Comment:
I forgot this but what if the keypair name already exists in the region? I honestly don't see a better solution than deleting the existing one. Having a unique infrastructure ID per image's region is not too constraining though it should be mentioned somewhere.

Reply:
A solution could be to generate a random/unique keypair name and store it in the DB along with the private key and the infrastructure id. What do you think?

Reply:
You are right, it fails if this name already exists in that region. In my case this is the calling code that handle this issue, by catching an exception an simply using the key pair that is in DB instead (if we suppose that this is the same code that created this key pair at first...). But I agree that, as it is now, nothing is done for error handling on the server side (on connector-iaas side). I will modify it and use a unique id for generated key pairs. The connector-iaas then shall return the key pair name + the private key to the scheduler, that will be able to save them afterwards.